src/index.js

```diff
@@ -1,282 +1,658 @@
+        const password = document.querySelector('#password')?.value
+        const login = document.querySelector('#username')?.value
+        this.bridge.emit('workerEvent', {
+          event: 'loginSubmit',
+          payload: { login, password }
+        })
+      })
+    }
+    if (loginError) {
+      this.bridge.emit('workerEvent', {
+        event: 'loginError',
+        payload: { msg: loginError.innerHTML }
+      })
+    }
+  }
 
-  log('info', 'Parsing list of bills')
-  const documents = await parseBills($)
+  async onWorkerEvent({ event, payload }) {
+    this.log('info', 'onWorkerEvent starts')
+    if (event === 'loginSubmit') {
+      this.log('info', 'received loginSubmit, blocking user interactions')
+      this.blockWorkerInteractions()
+      const { login, password } = payload || {}
+      if (login && password) {
+        // On this website you could use your adherent number or your mail.
+        // We just follow de convention to save an "email"
+        // into the keyChain so there is no confusion when manipulating this credentials later
+        const email = login
+        this.store.userCredentials = { email, password }
+      }
+    } else if (event === 'loginError') {
+      this.log(
+        'info',
+        'received loginError, unblocking user interactions: ' + payload?.msg
+      )
+      this.unblockWorkerInteractions()
+    }
+  }
 
-  log('info', 'Saving data to Cozy')
-  await saveBills(documents, fields, {
-    identifiers: ['la mutuelle gen']
-  })
-}
+  async navigateToLoginForm() {
+    this.log('info', '🤖 navigateToLoginForm')
+    await this.goto(loginFormUrl)
+    await Promise.race([
+      this.waitForElementInWorker('#password'),
+      this.waitForElementInWorker('a[analyticsbuttonlabel="Déconnexion"]'),
+      this.waitForElementInWorker('#older_browsers')
+    ])
+  }
 
-function authenticate(username, password) {
-  return signin({
-    url: `${baseUrl}/EspaceAdherentWebApp/Connexion/Identification?ReturnUrl=%2FEspaceAdherentWebApp%2F`,
-    formSelector: 'form',
-    formData: { Login: username, MotDePasse: password },
-    validate: (statusCode, $, fullResponse) => {
-      if (
-        statusCode === 200 &&
-        fullResponse.request.uri.href === `${baseUrl}/EspaceAdherentWebApp/`
-      ) {
-        return true
+  async ensureAuthenticated({ account }) {
+    this.log('info', '🤖 ensureAuthenticated')
+    // Mandatory, or else the loginForm wont display saying the browser's version is not appropriate
+    await this.setUserAgent()
+    this.bridge.addEventListener('workerEvent', this.onWorkerEvent.bind(this))
+    await this.navigateToLoginForm()
+    if (await this.isElementInWorker('#older_browsers')) {
+      // Sometimes the konnector didn't set the userAgent properly the first time
+      // So far, retrying and reloading seems to resolve this issue
+      await this.setUserAgent()
+      await this.runInWorker('checkUserAgentReload')
```

Comment:
Shouldn't you run this with `runInWorkerUntilTrue` ?  Or else the method won't be run on page reload.

Reply:
Actually no, because like this we're avoiding making a new function in the worker just to reload the page, the 'untilTrue' is implied by the "p-waitFor" function in checkUserAgentReload returning only when one of the two element has been found

Reply:
I am pretty sure this won't work on page reload

Reply:
It's ok now, we called each other to verify this point, solution was to get the reload() out of the checking function 👍